inst/include/dust/filter.hpp

```diff
@@ -72,6 +72,72 @@ std::vector<typename T::real_t> filter(Dust<T> * obj,
   return log_likelihood;
 }
 
+
+template <typename T>
+std::vector<typename T::real_t> filter_device(Dust<T> * obj,
+                                              filter_state_device<typename T::real_t>& state,
+                                              bool save_trajectories,
+                                              std::vector<size_t> step_snapshot) {
+  typedef typename T::real_t real_t;
+
+  const size_t n_particles = obj->n_particles();
+  const size_t n_data = obj->n_data();
+  const size_t n_pars = obj->n_pars_effective();
+
+  std::vector<real_t> ll_host(n_pars, 0);
+  dust::device_array<real_t> log_likelihood(ll_host);
+  dust::device_weights<real_t> weights(n_particles, n_pars);
+  dust::device_scan_state<real_t> scan;
+  scan.initialise(n_particles, weights.weights());
+
+  if (save_trajectories) {
+    state.trajectories.resize(obj->n_state(), n_particles, n_data);
+    obj->state(state.trajectories.values(), state.trajectories.value_offset());
```

Comment:
I see this is where state is used - this does look very similar to how the other state() interfaces look.

If we could do this as:

```
    obj->state(state.trajectories.values() + state.trajectories.value_offset());
```

I think it would be closer

Reply:
Leaving until #227 when this will change again